src/sqlfluff/rules/L003.py

```diff
@@ -318,96 +407,57 @@ def _eval(self, context: RuleContext) -> Optional[LintResult]:
-                # compare to them when assessing indent.
-                "hanging_lines": [],
-                # comment_lines keeps track of lines which are all comment.
-                "comment_lines": [],
-                # segments we've seen the last child of
-                "finished": set(),
-                # First non-whitespace node on a line.
-                "trigger": None,
-            }
-        else:
-            memory = context.memory
-
-        if context.segment.is_type("newline"):
-            memory["in_indent"] = True
-        elif memory["in_indent"]:
-            if context.segment.is_type("whitespace"):
-                # it's whitespace, carry on
-                pass
-            elif context.segment.segments or (
-                context.segment.is_meta
-                and context.segment.indent_val != 0  # type: ignore
-            ):
-                # it's not a raw segment or placeholder. Carry on.
-                pass
-            else:
-                memory["in_indent"] = False
-                # we're found a non-whitespace element. This is our trigger,
-                # which we'll handle after this if-statement
-                memory["trigger"] = context.segment
-        else:
-            # Not in indent and not a newline, don't trigger here.
-            pass
-
-        # Is this the last segment? If so, need to "flush" any leftovers.
-        is_last = self._is_last_segment(
-            context.segment, memory, context.parent_stack, context.siblings_post
+        is_ignorable = any(
+            el.is_type(*self._ignore_types) for el in context.parent_stack + (segment,)
         )
+        if is_ignorable:
+            return LintResult(memory=memory)
 
-        if not context.segment.is_type("newline") and not is_last:
-            # We only process complete lines or on the very last segment
-            # (since there may not be a newline on the very last line)..
+        if segment.is_type("newline"):
+            memory.in_indent = True
+        elif memory.in_indent:
+            is_non_raw_segment = bool(segment.segments)
+            is_placeholder = segment.is_meta and segment.indent_val != 0  # type: ignore
+            if not (segment.is_whitespace or is_non_raw_segment or is_placeholder):
+                memory.in_indent = False
+                # First non-whitespace element is our trigger
+                memory.trigger = segment
+
+        is_last = self.is_final_segment(context)
+        if not segment.is_type("newline") and not is_last:
+            # Process on line ends or file end
             return LintResult(memory=memory)
 
-        if raw_stack and raw_stack[-1] is not context.segment:
-            raw_stack = raw_stack + (context.segment,)
-        res = self._process_raw_stack(
-            raw_stack,
-            memory,
+        line_summaries = self._process_raw_stack(
+            raw_stack=raw_stack,
+            memory=memory,
```

Comment:
It should be possible to calculate all line on a single pass, (done in #2857 ) - I think this would improve performance further and simplify reasoning about the behaviour. (given a stream of tokens there is only one correct indent so we should be able to also avoid multiple passes by `self.crawl`

Reply:
Agreed that it should be possible to calculate all on a single pass. I don't think it's true to say there is only one correct indent however:

```
SELECT
    MAX(NULLIF(FN_X(
        a
     ))) AS this_in_good_indenting
FROM table
```

and

```
SELECT
    MAX(NULLIF(
        FN_X(a)
    )) AS this_is_also_good_indenting
FROM table
```

are both valid and different.

Reply:
Isn’t L003 only concerned with indents rather than adding/changing line breaks? So in that sense the above two examples are the same indents aren’t they?

Reply:
@alanmcruickshank  Is that a typo? the indentation looks the same on both of those. Maybe a space difference.

Not really sure about the arguments that both are valid - I feel if they are both valid, they should be supported as types of indentation as it would far simplify L003.

However thats a mute point cause L003 doesn't accept (which I think is a result of trying to accept too much).
```
SELECT 
    MAX(NULLIF(FN_X(
        a
    ))) AS this_in_good_indenting
FROM table
```
And worse still it forces this kind of thing
```
SELECT 
    FN(MAX(CASE
        WHEN TRUE
            THEN res
            END)) AS this_hurts_my_soul
FROM table
```

Reply:
@tunetheweb  L003 is the indentation one

Reply:
Oh I missed the fact all the closing brackets are indented slightly differently and thought @alanmcruickshank was more talking about the different line splits (which IMHO is not so much about indents and so nothing to do with L003).

I’m not sure I agree they are equally valid, and if we want to support the first (indenting closing brackets one space more) then that should be a configurable option so L003 works in either mode rather than trying to accept both.

Reply:
Ah yes - @tunetheweb is correct - I think I was making the point that given a stream of _code_ tokens there will be potentially multiple valid _whitespace_ configurations. That's because choosing where to split long lines is somewhat complicated.

If what you're actually saying is that given a stream of _code and newline_ tokens, that there is only one valid _indentation_ configuration (or at least only one _ideal_ setup), then **yes** I agree.

Reply:
L016 handles line splitting, and L003 only handle indents AFAIK.

Though not sure whether L016 also handles indents for any new splits, or leaves the next pass of L003 to deal with those.

Reply:
it's a bit of a mess right now. Hence #946.

Reply:
it's a bit of a mess right now. Hence https://github.com/sqlfluff/sqlfluff/issues/946.

Reply:
This Refac to L003 improves it. But one would need rewrite (and rule reconsideration to fix things overall).

And yes I was saying that given new lines and a stream we can calc the indents